### PR TITLE
[SC-64] freeze sdai market

### PR DESCRIPTION
**Pull request by hexonaut**
In preparation for disabling DAI borrows against sDAI we first need to freeze the market. Subsequent spell will reactivate with better params.

Goerli Spell: 0x2Ad00613A66D71Ff2B0607fB3C4632C47a50DADe  
Mainnet Spell: 0x843A0539Ca7466Abcb769f1c1d30C8423e13A297  

Tenderly Fork: https://dashboard.tenderly.co/shared/fork/simulation/8e9f6723-f12d-44d2-9997-a38fde9d82d8

### CODE REVIEW

File: src/proposals/20230712/SparkEthereum_20230712.t.sol [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+pragma solidity ^0.8.10;
+
+import 'forge-std/Test.sol';
+
+import { TestWithExecutor } from 'aave-helpers/GovHelpers.sol';
+import { IPool } from "aave-v3-core/contracts/interfaces/IPool.sol";
+import { IERC20WithPermit } from "aave-v3-core/contracts/interfaces/IERC20WithPermit.sol";
+
+import { SparkTestBase, ReserveConfig } from '../../SparkTestBase.sol';
+import { SparkEthereum_20230712 } from './SparkEthereum_20230712.sol';
```
Comment (lucas-manuel) [inline src/proposals/20230712/SparkEthereum_20230712.t.sol:11]:
```suggestion
import 'forge-std/Test.sol';

import { TestWithExecutor } from 'aave-helpers/GovHelpers.sol';

import { IERC20WithPermit } from "aave-v3-core/contracts/interfaces/IERC20WithPermit.sol";
import { IPool }            from "aave-v3-core/contracts/interfaces/IPool.sol";

import { SparkTestBase, ReserveConfig } from '../../SparkTestBase.sol';

import { SparkEthereum_20230712 } from './SparkEthereum_20230712.sol';
```

File: src/proposals/20230712/SparkEthereum_20230712.t.sol [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+contract SparkEthereum_20230712Test is SparkTestBase, TestWithExecutor {
+
+    SparkEthereum_20230712 internal payload;
+
+    address internal constant EXECUTOR = 0x3300f198988e4C9C63F75dF86De36421f06af8c4;
+    IPool internal constant POOL       = IPool(0xC13e21B648A5Ee794902342038FF3aDAB66BE987);
+    address internal constant SDAI     = 0x83F20F44975D03b1b09e64809B757c47f942BEeA;
```
Comment (lucas-manuel) [inline src/proposals/20230712/SparkEthereum_20230712.t.sol:19]:
```suggestion
    address internal constant EXECUTOR = 0x3300f198988e4C9C63F75dF86De36421f06af8c4;
    address internal constant SDAI     = 0x83F20F44975D03b1b09e64809B757c47f942BEeA;

    IPool internal constant POOL = IPool(0xC13e21B648A5Ee794902342038FF3aDAB66BE987);
```

File: src/proposals/20230712/SparkGoerli_20230712.sol [hunk 3]
```diff
@@ -0,0 +1,22 @@
+// SPDX-License-Identifier: AGPL-3.0
+pragma solidity ^0.8.10;
+
+import { SparkPayloadGoerli, IEngine, Rates, EngineFlags } from '../../SparkPayloadGoerli.sol';
+
+/**
+ * @title Freeze sDAI on Spark Goerli
+ * @author Phoenix Labs
+ * @dev This proposal freezes the sDAI market.
+ */
```
Comment (lucas-manuel) [inline src/proposals/20230712/SparkGoerli_20230712.sol:10]:
```suggestion
/**
 *  @title  Freeze sDAI on Spark Goerli
 *  @author Phoenix Labs
 *  @dev    This proposal freezes the sDAI market.
 */
```

File: src/proposals/20230712/SparkGoerli_20230712.t.sol [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+pragma solidity ^0.8.10;
+
+import 'forge-std/Test.sol';
+
+import { TestWithExecutor } from 'aave-helpers/GovHelpers.sol';
+import { IPool } from "aave-v3-core/contracts/interfaces/IPool.sol";
+import { IERC20WithPermit } from "aave-v3-core/contracts/interfaces/IERC20WithPermit.sol";
+
+import { SparkTestBase, ReserveConfig } from '../../SparkTestBase.sol';
+import { SparkGoerli_20230712 } from './SparkGoerli_20230712.sol';
```
Comment (lucas-manuel) [inline src/proposals/20230712/SparkGoerli_20230712.t.sol:11]:
```suggestion
import { TestWithExecutor } from 'aave-helpers/GovHelpers.sol';

import { IPool }            from "aave-v3-core/contracts/interfaces/IPool.sol";
import { IERC20WithPermit } from "aave-v3-core/contracts/interfaces/IERC20WithPermit.sol";

import { SparkTestBase, ReserveConfig } from '../../SparkTestBase.sol';

import { SparkGoerli_20230712 } from './SparkGoerli_20230712.sol';
```

File: src/proposals/20230712/SparkGoerli_20230712.t.sol [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+contract SparkGoerli_20230712Test is SparkTestBase, TestWithExecutor {
+
+    SparkGoerli_20230712 internal payload;
+
+    address internal constant EXECUTOR = 0x4e847915D8a9f2Ab0cDf2FC2FD0A30428F25665d;
+    IPool internal constant POOL       = IPool(0x26ca51Af4506DE7a6f0785D20CD776081a05fF6d);
+    address internal constant SDAI     = 0xD8134205b0328F5676aaeFb3B2a0DC15f4029d8C;
```
Comment (lucas-manuel) [inline src/proposals/20230712/SparkGoerli_20230712.t.sol:19]:
```suggestion
    address internal constant EXECUTOR = 0x4e847915D8a9f2Ab0cDf2FC2FD0A30428F25665d;
    address internal constant SDAI     = 0xD8134205b0328F5676aaeFb3B2a0DC15f4029d8C;

    IPool internal constant POOL = IPool(0x26ca51Af4506DE7a6f0785D20CD776081a05fF6d);
```

File: src/proposals/20230712/SparkGoerli_20230712.t.sol [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+        ReserveConfig[] memory configsBefore = _getReservesConfigs(POOL);
+        createConfigurationSnapshot('pre-Spark-Goerli-sDAI-Freeze', POOL);
+
+        _executePayload(address(payload));
+
+        ReserveConfig[] memory configsAfter = _getReservesConfigs(POOL);
+        createConfigurationSnapshot('post-Spark-Goerli-sDAI-Freeze', POOL);
+
+        assertEq(_findReserveConfig(configsBefore, SDAI).isFrozen, false);
+        assertEq(_findReserveConfig(configsAfter, SDAI).isFrozen, true);
```
Comment (lucas-manuel) [inline src/proposals/20230712/SparkGoerli_20230712.t.sol:39]:
```suggestion
        assertEq(_findReserveConfig(configsBefore, SDAI).isFrozen, false);
        assertEq(_findReserveConfig(configsAfter,  SDAI).isFrozen, true);
        
```

File: src/proposals/20230712/SparkGoerli_20230712.t.sol [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+        );
+    }
+
+    function testCantSupply() public {
+        _executePayload(address(payload));
+
+        deal(SDAI, address(this), 1 ether);
+        IERC20WithPermit(SDAI).approve(address(POOL), 1 ether);
+        vm.expectRevert(bytes('28'));      // Frozen
+        POOL.supply(SDAI, 1 ether, address(this), 0);
```
Comment (lucas-manuel) [inline src/proposals/20230712/SparkGoerli_20230712.t.sol:53]:
```suggestion
        
        vm.expectRevert(bytes('28'));  // Frozen
        POOL.supply(SDAI, 1 ether, address(this), 0);
```

File: src/proposals/20230712/SparkGoerli_20230712.t.sol [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+            SDAI
+        );
+    }
+
+    function testCantSupply() public {
+        _executePayload(address(payload));
+
+        deal(SDAI, address(this), 1 ether);
+        IERC20WithPermit(SDAI).approve(address(POOL), 1 ether);
+        vm.expectRevert(bytes('28'));      // Frozen
```
Comment (lucas-manuel) [inline src/proposals/20230712/SparkGoerli_20230712.t.sol:52]: Verified error code lgtm

File: src/proposals/20230712/SparkGoerli_20230712.t.sol [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+        createConfigurationSnapshot('pre-Spark-Goerli-sDAI-Freeze', POOL);
+
+        _executePayload(address(payload));
+
+        ReserveConfig[] memory configsAfter = _getReservesConfigs(POOL);
+        createConfigurationSnapshot('post-Spark-Goerli-sDAI-Freeze', POOL);
+
+        assertEq(_findReserveConfig(configsBefore, SDAI).isFrozen, false);
+        assertEq(_findReserveConfig(configsAfter, SDAI).isFrozen, true);
+        _noReservesConfigsChangesApartFrom(
```
Comment (lucas-manuel) [inline src/proposals/20230712/SparkGoerli_20230712.t.sol:40]: Nice

File: src/proposals/20230712/SparkEthereum_20230712.t.sol [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+        ReserveConfig[] memory configsAfter = _getReservesConfigs(POOL);
+        createConfigurationSnapshot('post-Spark-Ethereum-sDAI-Freeze', POOL);
+
+        assertEq(_findReserveConfig(configsBefore, SDAI).isFrozen, false);
+        assertEq(_findReserveConfig(configsAfter, SDAI).isFrozen, true);
+        _noReservesConfigsChangesApartFrom(
+            configsBefore,
+            configsAfter,
+            SDAI
+        );
```
Comment (lucas-manuel) [inline src/proposals/20230712/SparkEthereum_20230712.t.sol:44]:
```suggestion
        assertEq(_findReserveConfig(configsBefore, SDAI).isFrozen, false);
        assertEq(_findReserveConfig(configsAfter,  SDAI).isFrozen, true);

        _noReservesConfigsChangesApartFrom(
            configsBefore,
            configsAfter,
            SDAI
        );
```

**Review by lucas-manuel — CHANGES_REQUESTED** [review]
Reviewed spells and tests, lgtm just some formatting suggestions. Also will run a spell against goerli

**Review by amusingaxl — APPROVED** [review]
As a MakerDAO EA contributor, I confirm that the proposal executes the following actions:
- Freeze the `sDAI` market.

Checks made for both Goerli and Mainnet:

- [x] `sDAI` contract address match:
- [x] `EXECUTOR` is the spark SubDAO proxy
- [x] `_postExecute` hook freezes sDAI.

However, I should make it clear that I don't have enough knowledge about the inner workings of Spark to assess the impacts this spell would have in the system. I'm only assessing the impact this could have in the core MCD protocol.

File: src/proposals/20230712/SparkEthereum_20230712.sol [hunk 11]
```diff
@@ -0,0 +1,22 @@
+// SPDX-License-Identifier: AGPL-3.0
+pragma solidity ^0.8.10;
+
+import { SparkPayloadEthereum, IEngine, Rates, EngineFlags } from '../../SparkPayloadEthereum.sol';
+
+/**
+ * @title Freeze sDAI on Spark Ethereum
+ * @author Phoenix Labs
+ * @dev This proposal freezes the sDAI market.
+ */
```
Comment (lucas-manuel) [inline src/proposals/20230712/SparkEthereum_20230712.sol:10]:
```suggestion
/**
 *  @title  Freeze sDAI on Spark Ethereum
 *  @author Phoenix Labs
 *  @dev    This proposal freezes the sDAI market.
 */
```

**Review by lucas-manuel — COMMENTED** [review]
testCanWithdraw LGTM

**Comment by hexonaut** [comment]
Note Tenderly fork: https://dashboard.tenderly.co/phoenixlabs/sparklend/fork/47e5824a-f5bc-46df-96c9-4b12ffa54238/simulation/8e9f6723-f12d-44d2-9997-a38fde9d82d8

**Review by lucas-manuel — CHANGES_REQUESTED** [review]
Reviewed, changes LGTM.

**Review by lucas-manuel — APPROVED** [review]

**Pull request merged**